raiden_contracts/data/source/services/ServiceRegistry.sol

```diff
@@ -3,49 +3,167 @@ pragma solidity 0.5.4;
+            valid_till = now;
+        }
+        // Check against overflow.
+        require(valid_till < valid_till + 180 days);
+        valid_till = valid_till + 180 days;
+        assert(valid_till > service_valid_till[msg.sender]);
+        service_valid_till[msg.sender] = valid_till;
+
+        // Record the price
+        set_price = amount * 6 / 5;
+        set_price_at = now;
+
+        // Move the deposit in a new Deposit contract.
+        require(token.transferFrom(msg.sender, address(this), amount));
+        Deposit depo = new Deposit(address(token), valid_till, msg.sender);
+        require(token.transfer(address(depo), amount));
+
+        // Fire event
+        emit RegisteredService(msg.sender, valid_till, amount, depo);
     }
 
-    /// Set the URL used to access a service via HTTP.
-    /// When this is called for the first time, the service's ethereum address
-    /// is also added to `service_addresses`.
+    /// @notice Sets the URL used to access a service via HTTP.
+    /// Only a currently registered service can call this successfully.
+    /// @param new_url The new URL string to be stored.
     function setURL(string memory new_url) public {
+        require(now < service_valid_till[msg.sender]);
         require(bytes(new_url).length != 0, "new url is empty string");
-        if (bytes(urls[msg.sender]).length == 0) {
-            service_addresses.push(msg.sender);
-        }
         urls[msg.sender] = new_url;
     }
 
-    /// Returns number of registered services. Useful for accessing service_addresses.
-    function serviceCount() public view returns(uint) {
-        return service_addresses.length;
+    /// The amount of time till the price decreases to roughly 1/e.
+    uint constant decay_constant = 200 days; // Maybe make this configurable?
+
+    /// @notice Calculates the decreased price after a number of seconds.
+    /// @param _set_price The initial price.
+    /// @param _seconds_passed The number of seconds passed since the initial
+    /// price was set.
+    function decayed_price(uint256 _set_price, uint256 _seconds_passed) public
+        view returns (uint256) {
+        // We are here trying to approximate some exponential decay.
+        // exp(- X / A) where
+        //   X is the number of seconds since the last price change
+        //   A is the decay constant (A = 200 days correspnods to 0.5% decrease per day)
+
+        // exp(- X / A) ~~ P / Q where
+        //   P = 24 A^4
+        //   Q = 24 A^4 + 24 A^3X + 12 A^2X^2 + 4 AX^3 + X^4
+        // Note: swap P and Q, and then think about the Taylor expansion.
+
+        uint256 X = _seconds_passed;
+
+        if (X >= 2 ** 64) { // The computation below overflows.
+            return min_price;
+        }
+
+        uint256 A = decay_constant;
+
+        uint256 P = 24 * (A ** 4);
+        uint256 Q = P + 24*(A**3)*X + 12*(A**2)*(X**2) + 4*A*(X**3) + X**4;
+
+        uint256 price = _set_price * P / Q;
```

Comment:
Once you make `A` configurable, you might have to check for an overflow here as well.

Reply:
Yes, the next PR will contain lots of those checks.  I hope nobody needs the price decaying over billions of years.